pysmt/solvers/optimizer.py

```diff
@@ -0,0 +1,313 @@
+        optimizer can diverge (i.e. not terminate) if the objective is
+        unbounded (infinite or infinitesimal).
+        """
+        raise NotImplementedError
+
+    def _comparation_functions(self, objective_formula):
+        """Internal utility function to get the proper cast, LT and LE
+        function for the given objective formula
+        """
+        otype = self.environment.stc.get_type(objective_formula)
+        mgr = self.environment.formula_manager
+        if otype.is_int_type():
+            return mgr.Int, mgr.LT, mgr.LE
+        elif otype.is_real_type():
+            return mgr.Real, mgr.LT, mgr.LE
+        elif otype.is_bv_type():
+            return (lambda x: mgr.BV(x, otype.width)), mgr.BVULT, mgr.BVULE
+        else:
+            raise PysmtValueError("Invalid optimization function type: %s" % otype)
+
+
+class ExternalOptimizerMixin(Optimizer):
+    """An optimizer that uses an SMT-Solver externally"""
+
+    def _setup(self):
+        raise NotImplementedError
+
+    def _cleanup(self, client_data):
+        raise NotImplementedError
+
+    def _pareto_setup(self, client_data):
+        raise NotImplementedError
+
+    def _pareto_cleanup(self, client_data):
+        raise NotImplementedError
+
+    def _pareto_check_progress(self, client_data, cost_functions,
+                               costs_so_far, lts, les):
+        raise NotImplementedError
+
+    def _pareto_block_model(self, client_data, cost_functions, last_model, lts, les):
+        raise NotImplementedError
+
+    def _optimization_check_progress(self, client_data, cost_function,
+                                     cost_so_far, lt, le, strategy):
+        raise NotImplementedError
+
+    def _binary_bound(self, lb, ub):
+        if lb is None and ub is None:
+            return None
+        l,u = lb,ub
+        if lb is None and ub is not None:
+            l = ub - (abs(ub) + 1)
+        elif lb is not None and ub is None:
+            u = lb + abs(lb) + 1
+        return (l + u + 1) // 2
+
+    def _bound(self, lb, ub, strategy):
+        if strategy == 'binary':
+            return self._binary_bound(lb, ub)
+        if strategy == 'ub':
+            return ub
+        else:
+            raise PysmtValueError("Unknown optimization strategy '%s'" % strategy)
+
+    def optimize(self, cost_function, strategy='ub',
+                 feasible_solution_callback=None,
+                 step_size=1, **kwargs):
+        """This function performs the optimization as described in
+        `Optimizer.optimize()`. However. two additional parameters are
+        available:
+
+        `strategy` can be either 'binary' or 'ub'. 'binary' performs a
+        binary search to find the optimum, while 'ub' searches among
+        the satisfiable models.
+
+        `feasible_solution_callback` is a function with a single
+        argument or None. If specified, the function will be called
```

Comment:
The argument can be None? If so 
```suggestion
        argument (possibly None). If specified, the function will be called
```